build.sh

```diff
@@ -1,20 +1,161 @@
 #!/usr/bin/env bash
 
+# NOTE: Script for the CI builds. CI comes here from `.travis.yml`
+
+GHCJS_BUILD(){
+# NOTE: Function for GHCJS build that outputs its huge log into a file
+
+  # NOTE: Run the build into Log (log is too long for Travis)
+  "$@" &> "$ghcjsTmpLogFile"
+
+}
+
+SILENT(){
+# NOTE: Funtion that silences the build process
+# In normal mode outputs only the /nix/store paths
+
+  echo "Log: $ghcjsTmpLogFile"
+  # NOTE: Pass into the ghcjsbuild function the build command
+  if GHCJS_BUILD "$@"
+  then
+
+    # NOTE: Output log lines for stdout -> cachix caching
+    grep '^/nix/store/' "$ghcjsTmpLogFile"
+
+  else
+
+    # NOTE: Output log lines for stdout -> cachix caching
+    grep '^/nix/store/' "$ghcjsTmpLogFile"
+
+    # NOTE: Propagate the error state, fail the CI build
+    exit 1
+
+  fi
+
+}
+
+BUILD_PROJECT(){
 set -xe
 set -euo pipefail
 IFS=$'\n\t'
 
-GHCVERSION=${GHCVERSION:-ghc822}
-STRICT=${STRICT:-false}
-TRACING=${TRACING:-false}
-NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs/archive/6820e2f0dd16104961d6fc7e8e38846807159c4e.tar.gz
-
-if [ "$GHCVERSION" = "ghcjs" ]; then
-    nix-build --substituters 'https://nixcache.reflex-frp.org?trusted=1' ghcjs
-else
-    nix-build                                   \
-        --argstr compiler $GHCVERSION           \
-        --arg doTracing $TRACING                \
-        --arg doStrict $STRICT                  \
-        $@
+# NOTE: If var not imported - set to the default value
+GHCVERSION=${GHCVERSION:-ghc865}
+rev=${rev:-nixpkgs-unstable}
+NIX_PATH=${NIX_PATH:-"nixpkgs=https://github.com/nixos/nixpkgs/archive/$rev.tar.gz"}
+export NIX_PATH
+name=${name:-defaultBinaryName}
+pkgName=${pkgName:-defaultPkgName}
+failOnAllWarnings=${failOnAllWarnings:-'false'}
+checkUnusedPackages=${checkUnusedPackages:-'false'}
+doCoverage=${doCoverage:-'false'}
+doHaddock=${doHaddock:-'false'}
+doJailbreak=${doJailbreak:-'false'}
+doCheck=${doCheck:-'true'}
+doBenchmark=${doBenchmark:-'false'}
+enableExecutableProfiling=${enableExecutableProfiling:-'false'}
+enableLibraryProfiling=${enableLibraryProfiling:-'false'}
+buildFromSdist=${buildFromSdist:-'false'}
+buildStrictly=${buildStrictly:-'false'}
+disableOptimization=${disableOptimization:-'true'}
+buildStackProject=${buildStackProject:-'false'}
+# NOTE: *Oprparse* key is redifined in the code further
+generateOptparseApplicativeCompletion=${generateOptparseApplicativeCompletion:-'false'}
```

Comment:
Maybe just move this closer to the other code so the connection becomes obvious?

Reply:
Well, then we break the sectioning, currently, all env variables setup is at one place, I would just add -> directions to it.